rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalHttpServicePrebootSetup
+  extends Pick<HttpServiceSetup, 'auth' | 'csp' | 'basePath' | 'getServerInfo'> {
+  server: HttpServerSetup['server'];
+  externalUrl: ExternalUrlConfig;
+  registerRoutes(path: string, callback: (router: IRouter) => void): void;
+}
+```
+
+The only part of this contract that will be available to the preboot plugins via `CorePrebootSetup` is the API to register HTTP routes on the already running preboot HTTP server:
+
+```ts
+export interface HttpServicePrebootSetup {
+  registerRoutes(path: string, callback: (router: IRouter) => void): void;
+}
+```
+
+The Core HTTP context available to handlers of the routes registered on the preboot HTTP server will only expose the `uiSettings` service. As explained in the [UI Settings service section](#324-ui-settings-service), this service will give access to the **default** UI settings and their overrides set through Kibana configuration, if any.
+```ts
+// NOTE(azasypkin): the fact that the client is lazily initialized has been omitted for brevity.
+export interface PrebootCoreRouteHandlerContext {
+  readonly uiSettings: { client: IUiSettingsClient };
+}
+```
+
+The authentication and authorization components are not available at the `preboot` stage, and hence all preboot HTTP server routes can be freely accessed by anyone with access to the network Kibana is exposed to.
+
+Just as today, Kibana will shut the preboot HTTP server down as soon as it's ready to start the primary HTTP server.
+
+### 3.2.3 Elasticsearch service
+
+As mention in the [Motivation section](#2-motivation), the main goal of the interactive setup mode is to give the user a hassle-free way to configure Kibana connection to an Elasticsearch cluster. That means that users might provide certain connection information, and Kibana preboot plugins should be able to construct a new Elasticsearch client using this information to verify it and potentially call Elasticsearch APIs.
+
+To support this use case we'll add a new `preboot` method to the Elasticsearch service that will return the following contract, and make it available to the preboot plugins via `CorePrebootSetup`:
+
+```ts
+export interface ElasticsearchServicePrebootSetup {
+  readonly createClient: (
+    type: string,
+    clientConfig?: Partial<ElasticsearchClientConfig>
+  ) => ICustomClusterClient;
+}
+```
+
+The Elasticsearch clients created with `createClient` rely on the default Kibana Elasticsearch configuration and any configuration overrides specified by the consumer.
+
+__NOTE:__ We may need to expose a full or portion of Elasticsearch config to the preboot plugins for them to check if the user has already configured Elasticsearch connection. There are other ways to check that without direct access to the configuration though.
```

Comment:
> We may need to expose a full or portion of Elasticsearch config to the preboot plugins for them to check if the user has already configured Elasticsearch connection.

`Motivation` says we run `preboot` phase only if `it detects that user hasn't explicitly configured their own connection to Elasticsearch.` Does it mean that we run `preboot` when ES configuration is present, but Kibana cannot establish a network connection?

Reply:
That's a good question. 

> `Motivation` says we run `preboot` phase only if `it detects that user hasn't explicitly configured their own connection to Elasticsearch.`

The `preboot` will still run, but if interactive setup mode plugin detects that user has already configured connection to ES it 1) will not hold the boot, and 2) will instead render a "replacement" for the static `Kibana server is not ready yet` string (Design Team will help us to display something user-friendlier and that will automatically redirect user to the original URL when the primary server is up).

> Does it mean that we run `preboot` whenever ES configuration is present, but Kibana cannot establish a network connection?

Yes, if a user has already configured ES connection, we assume they know what they are doing no matter if ES is available or not. But it'd easy to improve experience in the future if we feel we need it.

Reply:
> I strongly feel we should be showing some limited status information on this UI to indicate what's happening inside Kibana. Historically, this information has required authentication to access (/api/status), but I think we could safely expose just the overall status level and summary text without requiring auth.

I agree. We may not have enough time to cover this in the initial implementation though, but I'll make sure we have everything in place to do this in the follow-up, if we want to.